amaranth/hdl/_nir.py

```diff
@@ -405,16 +412,81 @@ def add_value_cell(self, width: int, cell):
         cell_idx = self.add_cell(cell)
         return Value(Net.from_cell(cell_idx, bit) for bit in range(width))
 
-    def alloc_late_value(self, width: int):
-        self.last_late_net -= width
-        return Value(Net.from_late(self.last_late_net + bit) for bit in range(width))
+    def alloc_late_value(self, signal: _ast.Signal):
+        self.last_late_net -= len(signal)
+        value = Value(Net.from_late(self.last_late_net + bit) for bit in range(len(signal)))
+        for bit, net in enumerate(value):
+            self.late_to_signal[net] = signal, bit
+        return value
 
     @property
     def top(self):
         top = self.cells[0]
         assert isinstance(top, Top)
         return top
 
+    def check_comb_cycles(self):
+        class Cycle(Exception):
+            def __init__(self, start):
+                self.start = start
+                self.path = []
+
+        checked = set()
+        busy = set()
+
+        def traverse(net):
+            if net in checked:
+                return None
+
+            if net in busy:
+                raise Cycle(net)
+
+            busy.add(net)
+
+            try:
+                if net.is_const:
+                    pass
+                elif net.is_late:
+                    try:
+                        traverse(self.connections[net])
+                    except Cycle as e:
+                        sig, bit = self.late_to_signal[net]
+                        e.path.append((sig, bit, sig.src_loc))
+                        raise
+                else:
+                    for src, src_loc in self.cells[net.cell].comb_edges_to(net.bit):
+                        try:
+                            traverse(src)
+                        except Cycle as e:
+                            e.path.append((self.cells[net.cell], net.bit, src_loc))
+                            raise
```

Comment:
Please use a local return instead of a nonlocal return, since using a nonlocal return adds nothing.